### PR TITLE
fix(patterns): fix composing `AnyOf` and `AllOf` patterns

### PR DESCRIPTION
**Pull request by cpcloud**
- **test: add test for and-ing and or-ing allof/anyof**
- **fix(patterns): reference the correct attribute in sugar**

### CODE REVIEW

**Comment by cpcloud** [comment]
Seems like `inners` can't be accessed from Cython, the tests pass with pure Python.

**Comment by kszucs** [comment]
> Seems like `inners` can't be accessed from Cython, the tests pass with pure Python.

Right, you need to cast the object to the extension type first:

```py
return AnyOf(*self.inners, *other.inners)
```
 to

```py
left: AnyOf = cython.cast(AnyOf, self)
right: AnyOf = cython.cast(AnyOf, other)
return AnyOf(*left.inners, *right.inners)
```

**Comment by cpcloud** [comment]
Ah, roger that. Fixing it now.

File: koerce/patterns.py [hunk 1]
```diff
@@ -300,12 +296,8 @@ def __and__(self, other: Pattern) -> AllOf:
         -------
         New pattern that matches if both of the patterns match.
         """
-        if isinstance(self, AllOf) and isinstance(other, AllOf):
-            return AllOf(*self.patterns, *other.patterns)
-        elif isinstance(self, AllOf):
-            return AllOf(*self.patterns, other)
-        elif isinstance(other, AllOf):
-            return AllOf(self, *other.patterns)
```
Comment (cpcloud) [inline koerce/patterns.py:308]: I dunno, maybe the original but with casts is better, the entire implementation is in a single place and there's no method dispatching 🤔

**Review by kszucs — APPROVED** [review]
Sweet, thanks Phillip!

**Pull request merged**